### PR TITLE
Handle K8s jobs on job termination and K8s errors

### PR DESCRIPTION
**Pull request by Taadas**
## Summary & Motivation

Before this change, Dagster would not handle K8s jobs created using `execute_k8s_job` if Dagster job is terminated. For example, there is a job called `launch-job` that consists of operation called `launch-op` which calls `execute_k8s_job` method. `execute_k8s_job` creates a Kubernetes job named `launch-k8s`. There are two main scenarios when Dagster job `launch-job` will be terminated but Kubernetes job `launch-k8s` will keep running:
- If exception related to the Kubernetes is raised (e.g. `launch-k8s` timeouts) it means that Dagster will mark `lauch-op` as failed and terminate `launch-job`. However, `launch-k8s` Kubernetes job will keep running
- If pipeline execution was interrupted during the execution process (e.g. pipeline was manually cancelled). It means that Dagster will terminate `launch-job` and all of its running operations. However, `launch-k8s` Kubernetes job will keep running

Main issues with this behavior are:
- Observability. This really hurts the observability of the pipelines because it's hard to tell if the job was fully terminated
- Kubernetes load. It also can increase load on the Kubernetes cluster because there will be running jobs that are not needed anymore. Basically, computing resources are wasted
- Unexpected behavior. It can cause some unexpected behavior if the same Dagster job is launched again, because there could be two Kubernetes job running at the same time and doing the same thing (e.g. consuming from the same queue)

With this change, Dagster will handle Kubernetes jobs created using `execute_k8s_job` on Dagster job termination and Kubernetes errors.

## How I Tested These Changes

- `test_k8s_job_op_with_timeout_fail` test was modified to check if Kubernetes job created through `execute_k8s_job` method is deleted after Kubernetes related exception (particularly timeout in this case) is raised. `Timed out while waiting for pod to become ready` was changed to the `Timed out` because timeout can be raised in different places, not only when waiting for pod to become ready (`wait_for_pod` method) but also when waiting for job to complete (`wait_for_job` method) and so on
- `test_k8s_job_op_with_failure` test was modified because of two reasons:

  - Basically, it was doing the same thing as `test_k8s_job_op_with_timeout_fail` test, because timeout was set to 5 seconds and Kubernetes job was set to sleep for 10 seconds and then exit with non-zero exit code. This means that Kubernetes job will always timeout and will never exit with non-zero exit code. So, I removed timeout from the Kubernetes job and set it to exit with non-zero exit code immediately

  - Check if Kubernetes job created through `execute_k8s_job` method is deleted after Kubernetes related exception (particularly failed job in this case) is raised.
- `test_execute_k8s_job_terminate` test was added to check if Kubernetes job
created through `execute_k8s_job` method is deleted after Dagster job is
manually terminated

### CODE REVIEW

File: python_modules/libraries/dagster-k8s/dagster_k8s/ops/k8s_job_op.py [hunk 1]
```diff
@@ -345,36 +346,50 @@ def execute_k8s_job(
...
-        start_time=start_time,
-    )
+    try:
+        api_client.wait_for_job(
+            job_name=job_name,
+            namespace=namespace,
+            wait_timeout=timeout,
+            start_time=start_time,
+        )
+    except (DagsterExecutionInterruptedError, Exception) as e:
```
Comment (alangenfeld) [inline python_modules/libraries/dagster-k8s/dagster_k8s/ops/k8s_job_op.py:356]: instead of having this try/except block 3 times, should we just have one that cover from job creation until job success? Currently If we got an interrupt during the uncovered parts like log tailing we would not clean-up the job.

**Review by alangenfeld — COMMENTED** [review]
Wondering if we need an opt-out flag `execute_k8s_job` for this, strugling to imagine a scenario where someone wants the current behavior of letting orphaned jobs complete instead of stopping them.

File: python_modules/libraries/dagster-k8s/dagster_k8s/ops/k8s_job_op.py (hunk 1 above)
Comment (Taadas) [reply python_modules/libraries/dagster-k8s/dagster_k8s/ops/k8s_job_op.py]: Yeah, I agree with that, changed.

File: integration_tests/test_suites/k8s-test-suite/tests/test_integration.py [hunk 2]
```diff
@@ -24,6 +25,35 @@
...
+        try:
+            api_client.batch_api.read_namespaced_job_status(job_name, namespace)
+            return True
+        except kubernetes.client.rest.ApiException:
+            time.sleep(5)
+
+    return False
+
+
+def _wait_job_termination_availability(
```
Comment (gibsondan) [inline integration_tests/test_suites/k8s-test-suite/tests/test_integration.py:45]: nit: consider: _wait_until_job_can_be_terminated for readability

**Comment by gibsondan** [comment]
i think it's reasonable to assume that execute_k8s_job is running if and only if the underlying k8s job is running (i.e. i do not think we need an opt-out flag)

File: integration_tests/test_suites/k8s-test-suite/tests/test_integration.py (hunk 2 above)
Comment (Taadas) [reply integration_tests/test_suites/k8s-test-suite/tests/test_integration.py]: Changed.

File: python_modules/dagster-test/dagster_test/test_project/test_jobs/repo.py [hunk 3]
```diff
@@ -385,6 +386,23 @@ def slow_graph():
     slow_op()
 
 
+@op
+def execute_k8s_op(context):
```
[the remark this answers is not in the document]
Comment (Taadas) [reply python_modules/dagster-test/dagster_test/test_project/test_jobs/repo.py]: > which is, as i understand it, is to basically run for a long time so that it can be terminated
That's right, changed.

**Review by alangenfeld — APPROVED** [review]
thanks!
